x-pack/plugins/security_solution/server/endpoint/routes/actions/status.ts

```diff
@@ -52,77 +53,103 @@ export const actionStatusRequestHandler = function (
-          },
+    const response = await getPendingActions(esClient, agentIDs);
+
+    return res.ok({
+      body: {
+        data: response,
+      },
+    });
+  };
+};
+
+const getPendingActions = async (
+  esClient: ElasticsearchClient,
+  agentIDs: string[]
+): Promise<EndpointPendingActions[]> => {
+  // retrieve the unexpired actions for the given hosts
+
+  const recentActions = await searchUntilEmpty<EndpointAction>(esClient, {
+    index: AGENT_ACTIONS_INDEX,
+    body: {
+      query: {
+        bool: {
+          filter: [
+            { term: { type: 'INPUT_ACTION' } }, // actions that are directed at agent children
+            { term: { input_type: 'endpoint' } }, // filter for agent->endpoint actions
+            { range: { expiration: { gte: 'now' } } }, // that have not expired yet
+            { terms: { agents: agentIDs } }, // for the requested agent IDs
+          ],
         },
       },
-      {
-        ignore: [404],
-      }
-    );
-    const pendingActions =
-      recentActionResults.body?.hits?.hits?.map((a): EndpointAction => a._source!) || [];
+    },
+  });
 
-    // retrieve any responses to those action IDs from these agents
-    const actionIDs = pendingActions.map((a) => a.action_id);
-    const responseResults = await esClient.search<EndpointActionResponse>(
-      {
-        index: '.fleet-actions-results',
-        body: {
-          query: {
-            bool: {
-              filter: [
-                { terms: { action_id: actionIDs } }, // get results for these actions
-                { terms: { agent_id: agentIDs } }, // ignoring responses from agents we're not looking for
-              ],
-            },
-          },
+  // retrieve any responses to those action IDs from these agents
+  const actionIDs = recentActions.map((a) => a.action_id);
+  const responses = await searchUntilEmpty<EndpointActionResponse>(esClient, {
+    index: '.fleet-actions-results',
+    body: {
+      query: {
+        bool: {
+          filter: [
+            { terms: { action_id: actionIDs } }, // get results for these actions
+            { terms: { agent_id: agentIDs } }, // ignoring responses from agents we're not looking for
+          ],
         },
       },
+    },
+  });
+
+  // respond with action-count per agent
+  const pending: EndpointPendingActions[] = agentIDs.map((aid) => {
+    const responseIDsFromAgent = responses
+      .filter((r) => r.agent_id === aid)
+      .map((r) => r.action_id);
+    return {
+      agent_id: aid,
+      pending_actions: recentActions
+        .filter((a) => a.agents.includes(aid) && !responseIDsFromAgent.includes(a.action_id))
+        .map((a) => a.data.command)
+        .reduce((acc, cur) => {
+          if (cur in acc) {
+            acc[cur] += 1;
+          } else {
+            acc[cur] = 1;
+          }
+          return acc;
+        }, {} as EndpointPendingActions['pending_actions']),
+    };
+  });
+
+  return pending;
+};
+
+const searchUntilEmpty = async <T>(
```

Comment:
I like this helper utility. I can see it becoming a "common" utility for our APIs in the future if needed. There are several instances where we need to page through content and having something like this will avoid us having to duplicate code